Feat: #BBB-76 Elastic search를 활용한 서적 검색

platinouss opened the pull request:
## 작업 개요
키워드 기반 서적 검색 시, 기존에는 서적명 또는 저자 정보를 DB에서 풀스캔하여 탐색해야 했기 때문에 성능 개선을 위해 Elasticsearch 도입

- [x] Elasticsearch를 통한 서적 검색
- [x] NAVER Open API를 통해 조회된 서적 정보를 Elasticsearch에 추가 
- [x] 기술 서적 스터디에서 선정된 서적만 DB에 추가하도록 구성 

<!-- 작업에 대한 요약 설명을 남겨주세요. -->

## 전달 사항
### 1. 프로젝트에 적합한 의존성 추가
![image](https://github.com/Team-BomBomBom/Server/assets/70827921/bfd8ce4e-02ae-4a4f-8891-9b26794b76c4)
엘라스틱 서치의 경우 버전 마다 여러 기능이 추가되거나 수정되기 때문에, 공식 문서에 기재된 버전 정보를 참고했다.

현재 우리 프로젝트는 Spring Boot 3.3.0 (Spring Framework 6.1.8)을 사용 중이며, 호환되는 버전은 두 가지가 존재한다. 
엘라스틱 서치 8.13.4 버전을 사용한다면, 해당 버전에 호환되는 Spring Data Elasticsearch 5.3.1 버전으로 의존성을 추가했다.

### 2. 엘라스틱 서치 환경 구성
```Dockerfile
version: '3'
name: bombombom
services:
  es:
    image: docker.elastic.co/elasticsearch/elasticsearch:8.13.4
    environment:
      - node.name=es
      - discovery.type=single-node
      - discovery.seed_hosts=es
      - xpack.security.enabled=false
      - xpack.security.http.ssl.enabled=false
      - xpack.security.http.ssl.verification_mode=certificate
      - xpack.security.transport.ssl.enabled=false
      - xpack.security.transport.ssl.verification_mode=certificate
      - xpack.license.self_generated.type=basic
    volumes:
      - es-data:/usr/share/es/data
      - certs:/usr/share/elasticsearch/config/certs
    ports:
      - 9200:9200
  kibana:
    image: docker.elastic.co/kibana/kibana:8.13.4
    environment:
      - ELASTICSEARCH_HOSTS=http://es:9200
    ports:
      - 5601:5601
    depends_on:
      - es
volumes:
  certs:
    driver: local
  es-data:
    driver: local
```
docker-compose로 쉽게 구성할 수 있었다. (엘라스틱 서치에 담겨있는 데이터를 시각적으로 볼 수 있는 kibana도 설치했지만 현 시점에서는 엘라스틱 서치만 추가해도 괜찮을 것 같다)

엘라스틱 서치 8버전 부터는 https로 통신해야 하는 것이 default 값으로 설정되었다. 테스트이기도 하고, 실제 production 환경에서 클라이언트가 직접 엘라스틱 서치에 쿼리를 요청하는 경우가 없을 것이라고 판단하여 https 설정을 false로 설정하여 환경변수로 넘겨주었다.

만약 https 설정을 false로 하지 않는다면, 엘라스틱 서치에서 생성된 인증서를 JDK(또는 JRE) 내부의 인증서 저장소에 저장해줘야 스프링 내부에서 엘라스틱 서치로 https 통신을 할 수 있다. (기본적으로 Java의 경우 자체 인증서 저장소에서 해당 https 사이트의 인증 기관이 등록이 되지 않은 경우 에러를 반환하기 때문)
<!-- 작업과 관련된 전달 사항을 남겨주세요. -->

## 참고 자료
[Spring Data Elasticsearch](https://docs.spring.io/spring-data/elasticsearch/reference/elasticsearch/versions.html)
[Elasticsearch Logging](https://www.elastic.co/guide/en/elasticsearch/client/java-api-client/current/java-rest-low-usage-logging.html)
[bulk insert native query](https://www.elastic.co/guide/en/elasticsearch/reference/current/docs-bulk.html#docs-bulk-api-request)
[upsert native query](https://www.elastic.co/guide/en/elasticsearch/reference/current/docs-update.html)

<!-- 작업 시 참고했던 자료의 출처를 남겨주세요. -->

platinouss committed:
[BBB-76] ✨Feat: Elasticsearch를 활용한 서적 조회 기능 & bulk 연산으로 서적 정보 저장 기능 추가

NAVER Open API를 통해 조회된 서적 정보는 추후 서적 검색에 활용되기 위해 Elasticsearch에 저장되어야 합니다. 이때 bulk 연산으로 삽입되도록 구성했으며, 서적명으로 서적을 조회할 수 있도록 구성했습니다.

platinouss committed:
[BBB-76] 🛠️Refactor: 서적 조회 시 Elasticsearch에서 조회 하고, 실제 기술 서적 스터디에서 사용되는 서적 정보만 DB에 추가되도록 로직 수정

msjang4 reviewed (approved):
고생하셨습니다.

sseunghoon commented on build.gradle:
```diff
@@ -30,6 +30,7 @@ dependencies {
     implementation 'org.springframework.boot:spring-boot-starter-data-jpa'
     implementation 'org.springframework.boot:spring-boot-starter-security'
     implementation 'org.springframework.boot:spring-boot-starter-webflux'
+    implementation 'org.springframework.boot:spring-boot-starter-data-elasticsearch:3.3.1'
```
이거 숫자 안 넣어도 알아서 의존성 넣어줬던거 같은데 지우면 에러가 발생하나요?
말씀하셨던 버전 문제가 이건가요?

platinouss replied:
버전 정보를 기입하지 않으면 gradle이 알아서 적합한 버전을 주입해준다고 알고 있어서 큰 에러가 발생하지는 않을 것 같지만, gradle이 저희가 사용하는 엘라스틱 서치 버전까지는 알 수 없을 것 같아 명시하게 되었습니다. 관련 정보는 본문에 추가해두었습니다 !

sseunghoon reviewed (approved):
ES 도입 멋지네요 👍

platinouss committed:
[BBB-76] ✨Feat: DB에 서적 정보 추가 시 ES에 bookId 값 업데이트
스터디에 특정 기술 서적이 추가되면, 해당 서적 정보를 DB에 추가하고 해당 서적의 bookId 값을 Elasticsearch에도 추가시켜준다.

platinouss committed:
[BBB-76] ✨Feat: isbn 값으로 서적 조회 실패 시 exception 반환

서적 정보 DB에 추가 시 isbn 값으로 서적을 조회하게 되는데, 이때 Elasticsearch에 매핑되는 서적이 존재하지 않을 때 BookNotFoundException인 custom exception을 던지도록 구성

platinouss committed:
[BBB-76] ⚡️Fix: Elasticsearch에 동일한 서적이 중복으로 추가되는 이슈 해결

document의 id를 isbn으로 설정하지 않아서 동일한 서적이 중복으로 추가되는 이슈가 있었습니다. document의 id를 isbn으로 설정하여 해결했습니다.

platinouss committed:
[BBB-76] ⚡️Fix: Elasticsearch에 동일한 서적 추가 요청에도 서적 정보만 변경되도록 upsert 적용

NAVER Open API를 통해 조회된 서적 정보를 Elasticsearch에 추가 시, 모든 필드가 덮어씌워지게 되어 이미 bookId가 설정되어 있더라도 null로 업데이트되었습니다. 따라서 특정 필드만 추가되거나 업데이트되도록 upsert를 사용한 native 쿼리를 만들어 해결했습니다.

platinouss committed:
[BBB-76] ✅Test: 서적 검색 테스트 코드 추가 및 수정

Elasticsearch를 도입하면서, 서적 검색 테스트 시 수정되어야 할 부분과 새롭게 생성된 테스트 케이스를 추가했습니다.

platinouss committed:
[BBB-76] ⚙️Build: CI build 단계에 Elasticsearch 모듈 추가

platinouss committed:
[BBB-76] ✨Feat: Book -> BookResult 변환 메서드 추가

platinouss committed:
[BBB-76] 🛠️Refactor: 유사한 상위 custom exception을 상속하도록 수정

msjang4 commented on src/main/java/com/bombombom/devs/book/controller/dto/BookAddRequest.java:
```diff
@@ -1,15 +1,17 @@
 @Builder
 public record BookAddRequest(
-    @NotBlank(message = "공백일 수 없습니다.") String keyword
+    @NotBlank(message = "공백일 수 없습니다.") String isbn
```
의도적으로 Long대신 String으로  선언하신거 같은데 이유가 궁금합니다

platinouss replied:
아 이 부분 수정할게요 !

msjang4 commented on src/test/java/com/bombombom/devs/config/ElasticsearchTestConfig.java:
```diff
@@ -0,0 +1,44 @@
+    private static final GenericContainer<?> container;
+
+    static {
+        container = new GenericContainer<>(
+            new ImageFromDockerfile().withDockerfileFromBuilder(b -> {
+                b.from("docker.elastic.co/elasticsearch/elasticsearch:8.13.4").build();
+            })
+        ).withExposedPorts(9200)
+            .withEnv("node.name", "es-test")
+            .withEnv("discovery.type", "single-node")
+            .withEnv("xpack.security.enabled", "false")
+            .withEnv("xpack.security.http.ssl.enabled", "false")
+            .withEnv("xpack.security.http.ssl.verification_mode", "certificate")
+            .withEnv("xpack.security.transport.ssl.enabled", "false")
+            .withEnv("xpack.security.transport.ssl.verification_mode", "certificate")
+            .withEnv("xpack.license.self_generated.type", "basic");
+        container.start();
```
elastic search는 mysql과 다르게 github actions만으로는 Test가 불가능해서 TestConfig가 따로 필요한걸까요?

platinouss replied:
통합 테스트 시, 로컬 환경과 테스트 환경을 분리하기 위한 용도입니다 ! 위 config가 설정되면, 테스트를 위한 엘라스틱 서치 환경이 구성되고 해당 환경에서 통합 테스트를 진행할 수 있게 됩니당

platinouss committed:
[BBB-76] ⚡️Fix: DB에 저장되지 않은 서적만 추가하도록 수정

platinouss committed:
[BBB-76] ⚡️Fix: 서적 추가 요청 시 전달 받은 isbn 인자가 Long 타입으로 캐스팅되도록 수정
